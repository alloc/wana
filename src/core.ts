// Observation
export { o } from './o'
export { auto, Auto } from './auto'
export { when } from './when'
export { watch, Watcher } from './watch'
export { Observer, Change } from './observable'

// Escape hatches
export { no } from './no'
export { noto } from './noto'

// Internals
export { $O, $$ } from './symbols'
