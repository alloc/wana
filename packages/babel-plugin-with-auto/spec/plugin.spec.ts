import { transformSync } from '@babel/core'
import fs from 'fs'
import path from 'path'
import withAutoPlugin from '../src/plugin'

const fixture = fs.readFileSync(path.resolve(__dirname, 'fixture.tsx'), 'utf8')
forEachExec(fixture, /\n\/\/ ✂︎ ([^\n]+)\n/g, match => {
  const name = match[1]
  const code =
    `import { withAuto } from 'wana'\n` +
    fixture.slice(
      match.index + match[0].length,
      fixture.indexOf('\n// ✂︎', match.index + 1)
    )

  test(name, () => {
    const result = transformSync(code, {
      plugins: [
        // '@babel/plugin-syntax-jsx',
        ['@babel/plugin-syntax-typescript', { isTSX: true }],
        withAutoPlugin,
      ],
    })
    expect(result.code).toMatchSnapshot()
  })
})

function forEachExec(
  str: string,
  re: RegExp,
  each: (m: RegExpExecArray) => void,
  m?: RegExpExecArray | null
) {
  while ((m = re.exec(str))) each(m)
}
