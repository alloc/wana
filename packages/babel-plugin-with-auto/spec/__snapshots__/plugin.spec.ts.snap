// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`anonymous fn expression (no props) 1`] = `
"import { withAuto } from 'wana';
withAuto.dev(function ({
  $auto,
  $useCommit
}) {
  let $result,
      $commit = $useCommit($auto.start());

  $render: try {
    $result = <div />;
    break $render;
  } finally {
    $auto.stop();
  }

  return <>
    {$result}
    {$commit()}
  </>;
});"
`;

exports[`anonymous fn expression 1`] = `
"import { withAuto } from 'wana';
withAuto.dev(function ({
  $auto,
  $useCommit,
  ...props
}) {
  let $result,
      $commit = $useCommit($auto.start());

  $render: try {
    $result = <div {...props} />;
    break $render;
  } finally {
    $auto.stop();
  }

  return <>
    {$result}
    {$commit()}
  </>;
});"
`;

exports[`destructured props parameter 1`] = `
"import { withAuto } from 'wana';
withAuto.dev(({
  $auto,
  $useCommit,
  x
}) => {
  let $result,
      $commit = $useCommit($auto.start());

  $render: try {
    $result = <div {...x} />;
    break $render;
  } finally {
    $auto.stop();
  }

  return <>
    {$result}
    {$commit()}
  </>;
});"
`;

exports[`injected displayName 1`] = `
"import { withAuto } from 'wana';
const Foo = withAuto.dev(function Foo({
  $auto,
  $useCommit
}) {
  let $result,
      $commit = $useCommit($auto.start());

  $render: try {
    $result = <div />;
  } finally {
    $auto.stop();
  }

  return <>
    {$result}
    {$commit()}
  </>;
});
Foo.displayName = \\"Foo\\";"
`;

exports[`multi-line arrow fn (no props) 1`] = `
"import { withAuto } from 'wana';
withAuto.dev(({
  $auto,
  $useCommit
}) => {
  let $result,
      $commit = $useCommit($auto.start());

  $render: try {
    $result = <div />;
    break $render;
  } finally {
    $auto.stop();
  }

  return <>
    {$result}
    {$commit()}
  </>;
});"
`;

exports[`multi-line arrow fn 1`] = `
"import { withAuto } from 'wana';
withAuto.dev(({
  $auto,
  $useCommit,
  ...props
}) => {
  let $result,
      $commit = $useCommit($auto.start());

  $render: try {
    $result = <div {...props} />;
    break $render;
  } finally {
    $auto.stop();
  }

  return <>
    {$result}
    {$commit()}
  </>;
});"
`;

exports[`named fn expression (no props) 1`] = `
"import { withAuto } from 'wana';
withAuto.dev(function Foo({
  $auto,
  $useCommit
}) {
  let $result,
      $commit = $useCommit($auto.start());

  $render: try {
    $result = <div />;
    break $render;
  } finally {
    $auto.stop();
  }

  return <>
    {$result}
    {$commit()}
  </>;
});"
`;

exports[`named fn expression 1`] = `
"import { withAuto } from 'wana';
withAuto.dev(function Foo({
  $auto,
  $useCommit,
  ...props
}) {
  let $result,
      $commit = $useCommit($auto.start());

  $render: try {
    $result = <div {...props} />;
    break $render;
  } finally {
    $auto.stop();
  }

  return <>
    {$result}
    {$commit()}
  </>;
});"
`;

exports[`one-line arrow fn (no props) 1`] = `
"import { withAuto } from 'wana';
withAuto.dev(({
  $auto,
  $useCommit
}) => {
  let $result,
      $commit = $useCommit($auto.start());

  $render: try {
    $result = <div />;
  } finally {
    $auto.stop();
  }

  return <>
    {$result}
    {$commit()}
  </>;
});"
`;

exports[`one-line arrow fn 1`] = `
"import { withAuto } from 'wana';
withAuto.dev(({
  $auto,
  $useCommit,
  ...props
}) => {
  let $result,
      $commit = $useCommit($auto.start());

  $render: try {
    $result = <div {...props} />;
  } finally {
    $auto.stop();
  }

  return <>
    {$result}
    {$commit()}
  </>;
});"
`;

exports[`return inside if block (no else block) 1`] = `
"import { withAuto } from 'wana';
withAuto.dev(({
  $auto,
  $useCommit,
  ...props
}) => {
  let $result,
      $commit = $useCommit($auto.start());

  $render: try {
    if (props.xxx) {
      $result = <div {...props} />;
      break $render;
    }

    const {
      a,
      b
    } = props;
    $result = <div {...props} b={a} a={b} />;
    break $render;
  } finally {
    $auto.stop();
  }

  return <>
    {$result}
    {$commit()}
  </>;
});"
`;

exports[`return inside if block 1`] = `
"import { withAuto } from 'wana';
withAuto.dev(({
  $auto,
  $useCommit,
  ...props
}) => {
  let $result,
      $commit = $useCommit($auto.start());

  $render: try {
    if (props.xxx) {
      $result = <div {...props} />;
      break $render;
    } else {
      $result = <div {...props} xxy />;
      break $render;
    }
  } finally {
    $auto.stop();
  }

  return <>
    {$result}
    {$commit()}
  </>;
});"
`;

exports[`return inside nested function 1`] = `
"import { withAuto } from 'wana';
withAuto.dev(({
  $auto,
  $useCommit
}) => {
  let $result,
      $commit = $useCommit($auto.start());

  $render: try {
    $result = <div onClick={() => {
      return;
    }} />;
    break $render;
  } finally {
    $auto.stop();
  }

  return <>
    {$result}
    {$commit()}
  </>;
});"
`;
